src/parser.cpp

```diff
@@ -447,7 +452,8 @@ namespace ada::parser {
 
     // Let domain be the result of running UTF-8 decode without BOM on the percent-decoding of input.
     // Let asciiDomain be the result of running domain to ASCII with domain and false.
-    std::optional<std::string> ascii_domain = to_ascii(input, false);
+    // The most common case is an ASCII input, in which case we do not need to call the expensive 'to_ascii'.
+    std::optional<std::string> ascii_domain = input_is_ascii ? unicode::percent_decode(input) : to_ascii(input, false);
```

Comment:
I believe percent_decode might not be sufficient for this case.

Reply:
Yes. Because the percent_decode might produce non-ascii. Ah. Ok. It is a bit more complicated.

Reply:
I have figured it out. It is not clear that I have reached the ideal code, but it now passes all tests on my machine.